### PR TITLE
Parse the line table of GSYM files.

### PR DESCRIPTION
**Pull request by ThinkerYzu1**
The line table of GSYM file provide a way to translate an address to the file name and the line number of its source code if there is.

Signed-off-by: Kui-Feng Lee <kuifeng@fb.com>

### CODE REVIEW

File: src/gsym/types.rs [hunk 1]
```diff
@@ -14,6 +14,12 @@ pub struct Header {
     pub uuid: [u8; 20],
 }
 
+#[allow(dead_code)]
```
Comment (danielocfb) [inline src/gsym/types.rs:17]: Please stop adding dead code.

File: src/gsym.rs [hunk 2]
```diff
@@ -74,6 +77,93 @@ impl SymResolver for GsymResolver {
...
+                }
+            }
+
+            if row_cnt == 0 {
+                continue;
+            }
+
+            let finfo = self.ctx.file_info(lntab_row.file_idx as usize);
+            let dirname = self.ctx.get_str(finfo.directory as usize);
+            let filename = self.ctx.get_str(finfo.filename as usize);
```
Comment (danielocfb) [inline src/gsym.rs:152]: Who sanitized these offsets?

File: src/gsym.rs [hunk 3]
```diff
@@ -74,6 +77,93 @@ impl SymResolver for GsymResolver {
...
+                        break;
+                    }
+                }
+            }
+
+            if row_cnt == 0 {
+                continue;
+            }
+
+            let finfo = self.ctx.file_info(lntab_row.file_idx as usize);
```
Comment (danielocfb) [inline src/gsym.rs:150]: Who sanitized the index?

File: src/gsym.rs [hunk 4]
```diff
@@ -74,6 +77,93 @@ impl SymResolver for GsymResolver {
...
+    /// # Arguments
+    ///
+    /// * `addr` - A 64-bit unsigned integer representing the address to find the source code location for.
+    ///
+    /// # Returns
+    ///
+    /// `Some(AddressLineInfo)` if the location is found, otherwise `None`.
+    ///
+    fn find_line_info(&self, addr: u64) -> Option<AddressLineInfo> {
+        let addr = addr - self.loaded_address;
```
Comment (danielocfb) [inline src/gsym.rs:96]: Can underflow.

File: src/gsym/linetab.rs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        // DBG_FIRST_SPECIAL are considered special operators. These
+        // operators change both the line number and address of the
+        // virtual machine and emit a new row.
+        _ => {
+            let adjusted = (op - DBG_FIRST_SPECIAL) as i64;
+            // The range of line number delta is from min_delta to
+            // max_delta, including max_delta.
+            let range = ltbl_hdr.max_delta - ltbl_hdr.min_delta + 1;
+            let line_delta = ltbl_hdr.min_delta + (adjusted % range);
+            let addr_delta = adjusted / range;
```
Comment (danielocfb) [inline src/gsym/linetab.rs:107]: What if `range` is zero?

File: src/gsym.rs [hunk 6]
```diff
@@ -74,6 +77,93 @@ impl SymResolver for GsymResolver {
         None
     }
 
+    /// Finds the source code location for a given address.
+    ///
+    /// This function takes in an address and returns the file path,
+    /// line number and column of the line in the source code where
```
[the remark this answers is not in the document]
Comment (ThinkerYzu1) [reply src/gsym.rs:83]: Got it!

File: src/gsym.rs [hunk 7]
```diff
@@ -74,6 +77,93 @@ impl SymResolver for GsymResolver {
...
+    /// the address corresponds to. If it doesn't find any match it
+    /// returns None.
```
[the remark this answers is not in the document]
Comment (ThinkerYzu1) [reply src/gsym.rs:85]: Sure!

File: src/gsym/types.rs (hunk 1 above)
Comment (ThinkerYzu1) [reply src/gsym/types.rs:17]: I will remove this line.

File: src/gsym.rs [hunk 8]
```diff
@@ -83,11 +173,37 @@ impl SymResolver for GsymResolver {
...
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .join("data")
+            .join("test.gsym");
```
[the remark this answers is not in the document]
Comment (danielocfb) [reply src/gsym.rs:200]: > Do you mean to use CARGO_MANIFEST_DIR or alike?

Yes.

> This is a prebuilt binary. Generating gsym files need additional tools to be installed.

I am not sure I understand what this has to do with the topic at hand? I am not asking you to generate anything.

File: src/gsym/parser.rs [hunk 9]
```diff
@@ -111,8 +116,25 @@ impl<'a> GsymContext<'a> {
             ));
         }
         let addr_data_off_tab = &data[off..end_off];
-        let str_tab =
-            &data[strtab_offset as usize..(strtab_offset as usize + strtab_size as usize)];
+        off += num_addrs as usize * 4;
```
[the remark this answers is not in the document]
Comment (danielocfb) [reply src/gsym/parser.rs:119]: Yep, please introduce constants.

File: src/gsym.rs (hunk 2 above)
Comment (danielocfb) [reply src/gsym.rs:152]: > `get_str()` checks the boundary just like I mentioned in the earlier comment of `file_info()`.

[No](https://github.com/libbpf/blazesym/blob/e7e4327036631dea10a992b559ca39c789e842dc/src/gsym/parser.rs#L180)? That, by the way, is exactly why we should not introduce such APIs in the first place.

File: src/gsym.rs (hunk 3 above)
Comment (danielocfb) [reply src/gsym.rs:150]: Panicking is unacceptable error handling behavior.

File: src/gsym.rs (hunk 4 above)
Comment (danielocfb) [reply src/gsym.rs:96]: Please stop making inter-function assumptions like this. Even if this was the case, tomorrow we add a new caller and it's no longer ensured. It should be guarded against here where the operation is performed.

**Comment by danielocfb** [comment]
We spoke. I will finish up the pull request, because it is blocking upcoming work that relies on this functionality. I will close this one and create a new one once I have something.

**Pull request closed without merging**